### PR TITLE
Backed basic implementation

### PR DESCRIPTION
**Pull request by AdamAkiva**
**Start with the ESLint merge first.**  
A basic implementation of RESTful API for a user CRUD http based operations.  
You are familiar with the format of my writing so it shouldn't be that terrible.  
It is a large PR you I suggest separating it into parts and working on it when you have the time/energy.  
For any questions, add comments and I'll respond.  
Good luck, you'll need it.  
P.S I'm waiting for a PR from you for the FE, notify me when it's ready

### CODE REVIEW

File: be/src/entities/utils/validator.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  )
+    .strict()
+    .safeParse(obj);
+}
+
+export function checkAndParseErrors(
+  ...results: Zod.SafeParseReturnType<unknown, unknown>[]
+) {
+  const errs: Zod.ZodError<unknown>[] = [];
+  results.forEach((result) => {
```
Comment (Meital-Kozhidov) [inline be/src/entities/utils/validator.ts:48]: Isn't that just `filter`ing? why `foreach`?

File: be/src/types/index.ts
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/types/index.ts:1]: I'll think about it, I'm just a bit worried about circular dependencies (which was one of the reasons I used one file to begin with).
I'm more inclined towards a name change then file separation but I'll think about it

File: be/src/api-docs/openapi.yml [hunk 2]
```diff
@@ -0,0 +1,330 @@
+openapi: 3.0.3
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/api-docs/openapi.yml:1]: What extension for VSCode do you use for openAPI (aka swagger)?

File: be/src/db/schemas.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  uuid,
+  varchar
+} from 'drizzle-orm/pg-core';
+
+/**********************************************************************************/
+
+/**
+ * Notes:
+ * 1. All database related naming MUST be snake_case, this is the convention
+ * for SQL based databases. Otherwise the database quires need to be in quotation
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/db/schemas.ts:21]: Fixed

File: be/src/db/schemas.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        },
+        { onDelete: 'cascade', onUpdate: 'no action' }
+      ),
+    email: varchar('email', { length: 512 }).unique().notNull(),
+    password: varchar('password', { length: 256 }).notNull(),
+    // First delete of a user will be a soft-delete. Second delete will be a hard
+    // delete of that user.
+    // A read of soft-deleted user will work.
+    // A creation of a soft-deleted user will reactivate it.
+    // An update will be of a soft-deleted user will be forbidden.
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/db/schemas.ts:92]: Fixed

File: be/src/db/schemas.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+/**********************************************************************************/
+
+/**
+ * Notes:
+ * 1. All database related naming MUST be snake_case, this is the convention
+ * for SQL based databases. Otherwise the database quires need to be in quotation
+ * marks.
+ * 2. Primary key is uuid and not serial for two reasons: The first being,
+ * it will be must easier to replicate the database where every key is unique.
+ * The second being, keeping with the convention of the company. The overhead and
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/db/schemas.ts:25]: image
If it was intended to be a meme, if not I meant infinity.  
Considering it is a personal project it is irrelevant and I will remove it.  
(The bane of using templates is forgetting to remove the irrelevant parts)

File: be/src/server/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+export function handleMissedRoutes(req: Request, res: Response) {
+  return res.status(StatusCodes.NOT_FOUND).json(`'${req.url}' does not exist`);
+}
+
+// eslint-disable-next-line @typescript-eslint/max-params
+export function errorHandler(
+  err: Error,
+  _: Request,
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/middleware.ts:86]: I don't see a reason to make is so.  
Maybe if I had more than one variable ignored I would have named them.  
If it is ignored anyhow, I don't see why I should care how it's called

File: be/src/utils/config.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  });
+  if (missingValues) {
+    console.error(`\nMissing the following env vars:\n${missingValues}`);
+
+    process.exit(ERR_CODES.EXIT_NO_RESTART);
+  }
+};
+
+const checkMissingEnvVariables = (mode: Mode) => {
+  const errMap = new Map<string, string>([
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/utils/config.ts:62]: I don't understand what you meant here

File: be/src/entities/user/service/reactivate.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+      userInfoModel: userInfoModel,
+      userCredentialsModel: userCredentialsModel
+    },
+    userId: userId
+  });
+  if (!users.length) {
+    throw userNotFoundError(userId);
+  }
+  if (users[0].isActive) {
+    throw userAlreadyActive(userId);
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/entities/user/service/reactivate.ts:43]: Mistake from my end while not thinking ahead enough.
Already fixed a couple of days in the `general-fixes` branch

File: be/src/entities/user/service/reactivate.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+  return {
+    ...users[0],
+    isActive: true
+  };
+}
+
+/**********************************************************************************/
+
+async function checkUserStatus(params: {
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/entities/user/service/reactivate.ts:84]: The file no longer exists in future branches, so I'm ignoring this comment

File: be/src/entities/utils/validator.ts (hunk 1 above)
Comment (AdamAkiva) [reply be/src/entities/utils/validator.ts:48]: Filter return the elements which match the condition.  
Here, I don't use `result` directly, but I need the type narrowing and condition to create an array of `result.error`.
I could use filter & map (or reduce) instead if you think it will be more clear.
Another option could be:
```ts
const errs = results
    .map((result) => {
      if (!result.success) {
        return result;
      }

      return undefined;
    })
    .filter(filterNullAndUndefined);
```
But this uses 2 loops instead of 1.
If you have a thought about an implementation with 1 loop and cleaner logic, I'm all ears

Comment (AdamAkiva) [reply be/src/entities/utils/validator.ts:48]: Hell I may even just use the second option because I found it so cool and fuck performance

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/api-docs/openapi.yml:1]: Didn't have any, now installed this- 
image
Which left the earlier problem untouched, plus added another error on this line-
`        $ref: '#/components/requestBodies/ReactivateUser'`.

Is this now the right extension?

File: be/src/db/schemas.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/db/schemas.ts]: It was a meme, and also a note to notice that when copying your code

File: be/src/server/middleware.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/server/middleware.ts:86]: Well to me it's weird because of the type, meaning this `_: Request` looks a bit weird, but sure idc

File: be/src/entities/utils/validator.ts (hunk 1 above)
Comment (Meital-Kozhidov) [reply be/src/entities/utils/validator.ts]: I like the second option you presented the most too :) I assumed type narrowing was the issue

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/api-docs/openapi.yml:1]: One of, yes.
That line was a leftover which I forgot to remove, should be ok now

File: be/src/server/middleware.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/middleware.ts:86]: I'll keep it as is then

File: be/src/entities/user/route.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { Router, json } from '../../types/index.js';
+
+import * as Controller from './controller.js';
+
+/**********************************************************************************/
+
+const { createOne, readOne, updateOne, deleteOne, reactivateOne } = Controller;
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/entities/user/route.ts:7]: The idea behind it was naming conflicts.  
When imported this way, you can use the module name if multiple imports have the same.
With that said it is no longer relevant after some changes I've made.
Fixed

File: be/src/server/http.ts [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+  private readonly _db;
+  private readonly _logger;
+
+  private readonly _app;
+  private readonly _server;
+
+  public constructor(params: {
+    mode: Mode;
+    db: DatabaseHandler;
+    logger: Logger['handler'];
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/http.ts:48]: Logger is a type, what you're suggesting is invalid typescript syntax.
In typescript if you want to go into an type sub-field (from my knowledge) you can only do it with square brackets syntax

File: be/src/server/http.ts [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    // Disable 'x-powered-by' should be pretty clear. Reason to disable etag
+    // can be understood by this comprehensive answer: https://stackoverflow.com/a/67929691
+    this._app = express().disable('etag').disable('x-powered-by');
+    this._server = createServer(this._app);
+
+    this._attachConfigurations();
+    this._attachEventHandlers(this._logger);
+  }
+
+  public listen(port: number | string, callback?: () => void) {
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/http.ts:63]: Consider we receive it as env variable and env variable is always string (as a type), I allow both and just do a simple check.
This way it should work for all cases, regardless of docker or no docker.
As a side note, this is only ran once when the server is initialized, so I really don't care about performance or saving the condition in this case.

File: be/src/server/http.ts [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    this._server.close();
+  }
+
+  public async attachConfigurationMiddlewares(
+    allowedOrigins: Set<string> = new Set<string>()
+  ) {
+    this._app.use(
+      Middlewares.checkMethod(HttpServer.ALLOWED_METHODS),
+      cors({
+        credentials: true,
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/http.ts:79]: Good question, I was never sure. I will research and let you know

[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/server/http.ts:79]: **TLDR** Removed
For our project it is not needed.
For Infinity projects I'm not sure, from what I understand this allows sending cookies and auth header from cross-origin.
The frontend and backend have different origins if it is not staticly served (NGINX or w/e), so I don't if stuff will work without it.
To check it I'll need to serve locally on different origins with a project which supports it.
I couldn't bother or care less, so I'll just ignore it until (and if) I'll get an error about it.

File: be/src/types/index.ts
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/types/index.ts:1]: I'm keeping it as is for now, maybe on a later date I'll fix it

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/api-docs/openapi.yml:1]: You mean it was changed it a later branch? because i still see it

File: be/src/entities/user/route.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/entities/user/route.ts:7]: Didn't you push or is it fixed on a different branch?

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/api-docs/openapi.yml:1]: I have the same extension and have no such error **after** I've removed that line.
If you still have it, maybe you forgot to pull?

File: be/src/utils/config.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/utils/config.ts]: Fixed

File: be/src/entities/user/route.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/entities/user/route.ts]: Are you sure you pulled, because I see it in Github.
image
`be/base` branch, 1 hour ago (you commented this 30 minutes ago)

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/api-docs/openapi.yml:1]: image
Now I have this, this file hates me

File: be/src/entities/user/route.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/entities/user/route.ts]: I pulled now and it worked, weird.

File: be/src/api-docs/openapi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/api-docs/openapi.yml:1]: Let's try something else, if you copy the file content (as is) to [here](https://editor.swagger.io/) does it have any errors?

[the remark this answers is not in the document]
Comment (Meital-Kozhidov) [reply be/src/api-docs/openapi.yml:1]: I opened the project again, and now he doesn't shout on anything in this file, so it was probably just VS code acting up

File: be/__tests__/config/globalSetup.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  provide('urls', {
+    baseURL: `${serverEnv.base}:${serverEnv.port}/${serverEnv.apiRoute}`,
+    healthCheckURL: `${serverEnv.base}:${serverEnv.port}/${serverEnv.healthCheck.route}`
+  });
+  provide('db', {
+    name: `dashboard-pg-${mode}`,
+    url: dbUrl
+  });
+
+  const db = new DatabaseHandler({
```
Comment (Meital-Kozhidov) [inline be/__tests__/config/globalSetup.ts:30]: From here this is the same code you have in `main.ts`, why not reused?

File: be/__tests__/config/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import 'vitest';
+
+/**********************************************************************************/
+
+declare module 'vitest' {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+  export interface ProvidedContext {
```
Comment (Meital-Kozhidov) [inline be/__tests__/config/index.d.ts:7]: Where is this used?

File: be/__tests__/config/utils.ts [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+  console.error(
+    `Missing or invalid 'NODE_ENV' env value, should never happen.` +
+      ' Unresolvable, exiting...'
+  );
+
+  process.exit(ERR_CODES.EXIT_NO_RESTART);
+}
+
+function checkEnvVariables() {
```
Comment (Meital-Kozhidov) [inline be/__tests__/config/utils.ts:40]: This is the same as code you have in `src`

File: be/__tests__/config/globalSetup.ts (hunk 14 above)
Comment (AdamAkiva) [reply be/__tests__/config/globalSetup.ts:30]: The way I see it, the source code and the tests should be separated as much as possible.
I have imports of stuff from the source code, but they are mostly types and stuff for config or for unit tests.
In this case I don't think it fits, `main.ts` should have the configurations for the server startup.
Here the configurations for the test server is initialized.
Perhaps in this case they are the same, but in general this should be the all of the difference of why not to use the server same server instance for the tests.
**TLDR:** Not in this case

File: be/__tests__/config/utils.ts (hunk 16 above)
Comment (AdamAkiva) [reply be/__tests__/config/utils.ts:40]: The variables are different

File: be/__tests__/config/index.d.ts (hunk 15 above)
Comment (AdamAkiva) [reply be/__tests__/config/index.d.ts:7]: See the example here: https://vitest.dev/config/#globalsetup
Basically it will auto-complete the `inject` call from vitest

File: be/__tests__/config/globalSetup.ts (hunk 14 above)
Comment (Meital-Kozhidov) [reply be/__tests__/config/globalSetup.ts:30]: I still think it can utility function that gets params, but I get it

File: be/__tests__/integration/user.test.ts [hunk 17]
```diff
@@ -0,0 +1,1442 @@
...
+          expect(statusCode).toBe(StatusCodes.CREATED);
+          const userData = usersData.find((user) => {
+            return user.email === data.email;
+          })!;
+          expect(omit(data, 'id', 'createdAt', 'isActive')).toStrictEqual(
+            omit(userData, 'password')
+          );
+        });
+      });
+      it('A lot', async () => {
```
Comment (Meital-Kozhidov) [inline be/__tests__/integration/user.test.ts:101]: Same as `Multiple` with other number, why not separate to a function?

Comment (AdamAkiva) [reply be/__tests__/integration/user.test.ts:101]: The way I see and use tests is a complete stand-alone.
I don't want to rely on implementation of each other

File: be/__tests__/config/globalSetup.ts (hunk 14 above)
Comment (AdamAkiva) [reply be/__tests__/config/globalSetup.ts:30]: Thank you.
Unless I see another approach with I like more, I'll keep it as is

File: be/__tests__/integration/user.test.ts (hunk 17 above)
Comment (Meital-Kozhidov) [reply be/__tests__/integration/user.test.ts:101]: Thought you'd say this

File: be/__tests__/config/utils.ts (hunk 16 above)
Comment (Meital-Kozhidov) [reply be/__tests__/config/utils.ts:40]: Again, could be a function, but I get you don't the test & src correlation so sure

Comment (AdamAkiva) [reply be/__tests__/config/utils.ts:40]: Thank you

File: be/__tests__/integration/user.test.ts (hunk 17 above)
Comment (AdamAkiva) [reply be/__tests__/integration/user.test.ts:101]: I would say surprise!
But you read my previous answers, it's the same approach unless I find a good reason to change it

File: be/__tests__/utils.ts [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+    response: createResponse<Response>(resOptions)
+  };
+}
+
+export async function stressTest(url: string, options: StressTestOptions) {
+  return await new Promise<autocannon.Result>((resolve, reject) => {
+    // Pay attention when changing these value.
+    // The tcp sockets are reused by autocannon which may cause an infinite
+    // timeout, if not configured correctly. Match these values to the
+    // `maxRequestsPerSocket` config of the http server
```
Comment (Meital-Kozhidov) [inline be/__tests__/utils.ts:185]: How is it matched?

Comment (AdamAkiva) [reply be/__tests__/utils.ts:185]: You are correct, the terminology is misleading.
I meant that if you change one you need to closely watch if you need to change the other as well.
Fixed the comment

**Review by Meital-Kozhidov — APPROVED** [review]
